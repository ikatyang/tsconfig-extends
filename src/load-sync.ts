import * as fs from 'fs';
import * as path from 'path';

export interface TSConfig {
  extends?: string;
  compilerOptions?: CompilerOptions;
}

export interface CompilerOptions {
  [key: string]: any;
}

export function load_file_sync(filename: string, cwd: string = process.cwd()): CompilerOptions {
  const _filename = path.resolve(cwd, filename);
  const json = _load_json_sync(_filename);
  return load_config_sync(json, path.dirname(_filename));
}

export function load_config_sync(tsconfig: TSConfig, cwd: string = process.cwd()): CompilerOptions {
  return _load_config_sync(tsconfig, cwd);
}

function _load_json_sync(filename: string): TSConfig {
  const content = fs.readFileSync(filename, 'utf8');
  return /^\s*$/.test(content)
    ? {}
    : JSON.parse(content);
}

function _load_config_sync(tsconfig: TSConfig, cwd: string, stack: string[] = []): CompilerOptions {
  const {
    compilerOptions: compiler_options = {},
  } = tsconfig;

  if (tsconfig.extends === undefined) {
    return compiler_options;
  }

  const parent_filename = path.resolve(cwd, tsconfig.extends);
  if (stack.indexOf(parent_filename) !== -1) {
    throw new Error(`Circular dependency detected: ${stack.join(' -> ')}`);
  }

  const {
    extends: parent_extends,
    compilerOptions: parent_compiler_options,
  } = _load_json_sync(parent_filename);

  return _load_config_sync(
    {
      extends: parent_extends,
      compilerOptions: {
        ...parent_compiler_options,
        ...compiler_options,
      },
    },
    path.dirname(parent_filename),
    [...stack, parent_filename],
  );
}
