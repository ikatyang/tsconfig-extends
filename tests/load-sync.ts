import { load_config_sync, load_file_sync } from '../src/index';

const load_fixture = (name: string) =>
  load_file_sync(`./fixtures/${name}/tsconfig.json`);

const load_fixture_cwd = (name: string) =>
  load_file_sync(`../fixtures/${name}/tsconfig.json`, __dirname);

it('should return correclty (general)', () => {
  expect(load_fixture_cwd('general')).toMatchSnapshot();
});

it('should return correclty (empty)', () => {
  expect(load_fixture_cwd('empty')).toMatchSnapshot();
});

it('should return correctly (extends)', () => {
  expect(load_fixture_cwd('extends')).toMatchSnapshot();
});

it('should throw error (circular)', () => {
  expect(() => load_fixture_cwd('circular')).toThrowError();
});

it('should return correctly with filename and default cwd', () => {
  expect(load_fixture('general')).toMatchSnapshot();
});

it('should return correctly with tsconfig and default cwd', () => {
  expect(
    load_config_sync({
      extends: './fixtures/general/tsconfig.json',
      compilerOptions: {
        test: true,
      },
    }),
  ).toMatchSnapshot();
});
