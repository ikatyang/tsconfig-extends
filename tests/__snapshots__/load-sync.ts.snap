// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correclty (empty) 1`] = `Object {}`;

exports[`should return correclty (general) 1`] = `
Object {
  "test:general": true,
}
`;

exports[`should return correctly (extends) 1`] = `
Object {
  "test:extends": 0,
  "test:extends:0": true,
  "test:extends:1": true,
  "test:extends:2": true,
}
`;

exports[`should return correctly with filename and default cwd 1`] = `
Object {
  "test:general": true,
}
`;

exports[`should return correctly with tsconfig and default cwd 1`] = `
Object {
  "test": true,
  "test:general": true,
}
`;
